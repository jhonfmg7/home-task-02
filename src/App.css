body {
    box-sizing: border-box;
    font-family: 'Roboto', sans-serif;
    font-weight: 400;
}

*,
*::before,
*::after {
    box-sizing: inherit;
}

.container {
    text-align: center;
}

.button {
    margin: 0 10px;
    border-radius: 50%;
    height: 50px;
    width: 50px;
    font-size: 24px;
}