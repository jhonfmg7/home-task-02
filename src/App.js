import React from "react";

// Components
import Counter from "./components/Counter.tsx";

function App() {
  return (
    <div className="container">
      <h1>Hello World!</h1>
      <Counter />
    </div>
  );
}

export default App;
