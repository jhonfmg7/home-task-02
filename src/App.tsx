import * as React from 'react';
import { Provider } from 'react-redux';

// Interface
import Movie from './types/movie.interface';

// Store
import store from './redux/store';

// Context
import AppContext from './context/AppContext';

// Components
import Header from './components/header';
import Main from './components/main';
import Footer from './components/footer';
import ErrorBoundary from './components/ErrorBoundary';

function App() {
  // Local State
  const [movieSelected, setMovieSelected] = React.useState<Movie | null>(null);
  const [isOpenMoreInfoModal, setIsOpenMoreInfoModal] = React.useState(false);

  return (
    <Provider store={ store }>
      <div className="App" data-testid="app">
          <ErrorBoundary>
              <AppContext.Provider value={{ movieSelected, setMovieSelected, isOpenMoreInfoModal, setIsOpenMoreInfoModal }}>
                <Header />
                <Main />
                <Footer />
              </AppContext.Provider>
          </ErrorBoundary>
      </div>
    </Provider>
  );
}

export default App;
