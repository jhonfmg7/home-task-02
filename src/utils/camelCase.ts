const camelCase = (str: string) => {
    return str[0].toUpperCase() + str.slice(1, str.length + 1)
}

export default camelCase;