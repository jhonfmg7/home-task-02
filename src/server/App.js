import { hydrate } from "react-dom";
import App from "../App.tsx";

const MyApp = hydrate(App, document.getElementById('app'));

export default MyApp;