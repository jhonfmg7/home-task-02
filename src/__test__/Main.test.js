import * as React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

// Store
import store from '../redux/store';

// components
import App from '../App';
import Main from '../components/main';

// Utils
import camelCase from '../utils/camelCase';

afterEach( cleanup );

test('<Main /> validate UI and movies showed at main section', () => {
    const { getByTestId } = render(
        <Provider store={ store }>
            <BrowserRouter>
                <Main />
            </BrowserRouter>
        </Provider>
    );
    
    const mainContainer = getByTestId('main');
    expect( mainContainer ).toBeInTheDocument();
    expect( mainContainer.tagName ).toBe('MAIN');
});

test('<Navbar /> validate when change genre or sortby successful', () => {
    const { getByTestId } = render(
        <Provider store={ store }>
            <BrowserRouter>
                <App />
            </BrowserRouter>
        </Provider>
    );

    const homeLink = getByTestId('home_link');
    expect( homeLink ).toBeInTheDocument();
    fireEvent.click( homeLink, { button: 0 } );

    let params = new URLSearchParams(window.location.search);

    // Genres
    const genreOptionAll = getByTestId('genre_option_all');
    expect( genreOptionAll ).toBeInTheDocument();
    expect( genreOptionAll.textContent ).toBe('All');

    fireEvent.click(genreOptionAll);
    params = new URLSearchParams(window.location.search);
    expect( camelCase(params.get('genre')) ).toBe(genreOptionAll.textContent);

    const genreOptionAction = getByTestId('genre_option_action');
    expect( genreOptionAction ).toBeInTheDocument();
    expect( genreOptionAction.textContent ).toBe('Action');

    fireEvent.click(genreOptionAction);
    params = new URLSearchParams(window.location.search);
    expect( camelCase(params.get('genre')) ).toBe(genreOptionAction.textContent);

    const genreOptionDocumentary = getByTestId('genre_option_documentary');
    expect( genreOptionDocumentary ).toBeInTheDocument();
    expect( genreOptionDocumentary.textContent ).toBe('Documentary');

    fireEvent.click(genreOptionDocumentary);
    params = new URLSearchParams(window.location.search);
    expect( camelCase(params.get('genre')) ).toBe(genreOptionDocumentary.textContent);

    const genreOptionComedy = getByTestId('genre_option_comedy');
    expect( genreOptionComedy ).toBeInTheDocument();
    expect( genreOptionComedy.textContent ).toBe('Comedy');

    fireEvent.click(genreOptionComedy);
    params = new URLSearchParams(window.location.search);
    expect( camelCase(params.get('genre')) ).toBe(genreOptionComedy.textContent);

    const genreOptionHorror = getByTestId('genre_option_horror');
    expect( genreOptionHorror ).toBeInTheDocument();
    expect( genreOptionHorror.textContent ).toBe('Horror');

    fireEvent.click(genreOptionHorror);
    params = new URLSearchParams(window.location.search);
    expect( camelCase(params.get('genre')) ).toBe(genreOptionHorror.textContent);

    const genreOptionCrime = getByTestId('genre_option_crime');
    expect( genreOptionCrime ).toBeInTheDocument();
    expect( genreOptionCrime.textContent ).toBe('Crime');

    fireEvent.click(genreOptionCrime);
    params = new URLSearchParams(window.location.search);
    expect( camelCase(params.get('genre')) ).toBe(genreOptionCrime.textContent);

    // SortBy
    const sortBySelect = getByTestId('sortby_select');
    expect( sortBySelect ).toBeInTheDocument();

    fireEvent.change(sortBySelect, { target: { value: 'release_date' } });
    expect( sortBySelect.value ).toBe('release_date');
    expect( sortBySelect.value ).not.toBe('vote_average');
    expect( sortBySelect.value ).not.toBe('runtime');

    params = new URLSearchParams(window.location.search);
    expect( params.get('sortBy') ).toBe('release_date');

    fireEvent.change(sortBySelect, { target: { value: 'vote_average' } });
    expect( sortBySelect.value ).toBe('vote_average');
    expect( sortBySelect.value ).not.toBe('release_date');
    expect( sortBySelect.value ).not.toBe('runtime');

    params = new URLSearchParams(window.location.search);
    expect( params.get('sortBy') ).toBe('vote_average');

    fireEvent.change(sortBySelect, { target: { value: 'runtime' } });
    expect( sortBySelect.value ).toBe('runtime');
    expect( sortBySelect.value ).not.toBe('vote_average');
    expect( sortBySelect.value ).not.toBe('release_date');

    params = new URLSearchParams(window.location.search);
    expect( params.get('sortBy') ).toBe('runtime');
});

test('<Main /> validate each card movie exists and is successful showed', async() => {
    const { findAllByTestId, queryAllByTestId } = render(
        <Provider store={store}>
            <BrowserRouter>
                <Main />
            </BrowserRouter>
        </Provider>
    );

    const cardMovie = await findAllByTestId('card_movie');
    expect( cardMovie.length ).toBe(9)
    expect( cardMovie ).toMatchSnapshot();

    // Movie Menu Open
    const movieMenu = await queryAllByTestId('movie_menu');
    expect( movieMenu.length ).not.toBe(1);

    const movieMenuButton = await findAllByTestId('movie_menu_button');
    expect( movieMenuButton.length ).toBe(9);

    const movieMenuCloseButton = await queryAllByTestId('movie_menu_close_button');
    expect( movieMenuCloseButton.length ).toBe(0);

    fireEvent.click( movieMenuButton[0] );
    const movieMenuTwo = await queryAllByTestId('movie_menu');
    expect( movieMenuTwo.length ).toBe(1);

    const movieMenuCloseButtonTwo = await findAllByTestId('movie_menu_close_button');
    expect( movieMenuCloseButtonTwo.length ).toBe(1);
    expect( movieMenuCloseButtonTwo[0].textContent ).toBe('x');
    expect( movieMenuCloseButtonTwo[0].tagName ).toBe('H2');

    // Close and open again movie menu
    fireEvent.click( movieMenuCloseButtonTwo[0] );
    expect( movieMenu.length ).not.toBe(1);
    fireEvent.click( movieMenuButton[0] );

    // Delete Modal Open
    const deleteModal = await queryAllByTestId('delete_modal');
    expect( deleteModal.length ).toBe(0);
    expect( deleteModal.length ).not.toBe(1);

    const movieMenuButtonsContainer = await findAllByTestId('movie_menu_buttons_container');
    expect( movieMenuButtonsContainer[0].tagName ).toBe('UL');

    const deleteModalButton = await findAllByTestId('delete_modal_button');
    expect( deleteModalButton[0].textContent ).toBe("Delete");
    expect( deleteModalButton[0].tagName ).toBe("LI");

    const editModalButton = await findAllByTestId('edit_modal_button');
    expect( editModalButton[0].textContent ).toBe("Edit");
    expect( editModalButton[0].tagName ).toBe("LI");

    fireEvent.click(deleteModalButton[0]);
    const deleteModalTwo = await findAllByTestId('delete_modal');
    expect( deleteModalTwo.length ).toBe(1);

    const deleteModalTitle = await findAllByTestId('delete_modal_title');
    expect( deleteModalTitle[0].tagName ).toBe('P');
    expect( deleteModalTitle[0].textContent ).toBe('Are you sure you want to delete this movie?');

    const deleteModalConfirmButton = await findAllByTestId('delete_modal_confirm_button');
    expect( deleteModalConfirmButton[0].tagName ).toBe('BUTTON');
    expect( deleteModalConfirmButton[0].textContent ).toBe('Confirm');

    // Delete Modal Close
    const modalCloseButton = await findAllByTestId('modal_close_button');
    expect( modalCloseButton[0].tagName ).toBe('H2');
    expect( modalCloseButton[0].textContent ).toBe('X');

    fireEvent.click( modalCloseButton[0] );
    expect( deleteModal.length ).not.toBe(1);
    expect( deleteModalTwo[0] ).toMatchObject({});

    // Movie Menu Close
    fireEvent.click( movieMenuButton[0] );
    expect( movieMenu.length ).not.toBe(1);
    expect( movieMenuTwo[0] ).toMatchObject({});
});

test('<AddOrEditForm /> validate reset form funcionality', async() => {
    const { getByTestId, findAllByTestId, queryByTestId } = render(
        <Provider store={store}>
            <BrowserRouter>
                <App />
            </BrowserRouter>
        </Provider>
    );

    const addMovieButton = getByTestId('add_movie_button');
    fireEvent.click(addMovieButton);

    const resetFormMovieButton = getByTestId('new_movie_reset_form_button');
    expect( resetFormMovieButton ).toBeInTheDocument();

    const newMovieInputTitle = getByTestId('new_movie_input_title');
    expect( newMovieInputTitle ).toBeInTheDocument();
    expect( newMovieInputTitle.value ).toBe('');
    
    const newMovieSelectGenre = getByTestId('new_movie_input_genres');
    expect( newMovieSelectGenre ).toBeInTheDocument();
    expect( newMovieSelectGenre.value ).toBe('');
    expect( newMovieSelectGenre.className ).toBe('modal-module__input___3hdhz');
    
    // Validate Class of inputs not error
    expect( newMovieInputTitle.className ).toBe('modal-module__input___3hdhz');
    const newMovieSubmitButton = getByTestId('new_movie_submit_button');

    expect( newMovieSubmitButton ).toBeInTheDocument();
    await act(() => {
        fireEvent.click(newMovieSubmitButton);
    });

    // Validate Class of inputs error
    expect( newMovieInputTitle.className ).toBe('modal-module__inputWithError___3zQMu');
    expect( newMovieInputTitle.className ).not.toBe('modal-module__input___3hdhz');
    expect( newMovieSelectGenre.className ).toBe('modal-module__inputWithError___3zQMu');
    expect( newMovieSelectGenre.className ).not.toBe('modal-module__input___3hdhz');

    await act(() => {
        fireEvent.change(newMovieInputTitle, { target: { value: 'Testing title' } })
    });
    expect( newMovieInputTitle.value ).toBe('Testing title');

    // Reset form
    await act(() => {
        fireEvent.click(resetFormMovieButton);
    });
    expect( newMovieInputTitle.value ).toBe('');
})

