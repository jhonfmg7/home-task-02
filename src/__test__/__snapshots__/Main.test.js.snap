// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> validate each card movie exists and is successful showed 1`] = `
[
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Guardians of the Galaxy Vol. 3"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Guardians of the Galaxy Vol. 3
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2020
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Action & Adventure & Science Fiction
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Transformers 7"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Transformers 7
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2019
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Science Fiction & Action & Adventure
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Untitled Avengers"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/aMETsaNNcDc6g5ZatQtVbySnSaA.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Untitled Avengers
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2019
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Action & Adventure & Science Fiction
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Shazam!"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/yUOJHa9XmB1H0iYodG9Kb3YCc9T.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Shazam!
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2019
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Action & Adventure & Fantasy & Science Fiction
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="X-Men: Dark Phoenix"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/6qmsupE0opYPIaBGe7T5D2FBzLs.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          X-Men: Dark Phoenix
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2019
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Action & Science Fiction
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Journey 3: From the Earth to the Moon"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/98tbNloMBztAVnWpAznKKVUdi2O.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Journey 3: From the Earth to the Moon
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2018
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Action & Adventure & Fantasy & Science Fiction
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Aquaman"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/sshNRCbtU2ON8SiWLycQzr05doX.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Aquaman
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2018
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Action & Fantasy & Science Fiction
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Mission: Impossible - Fallout"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/A2W75CVz9l8gEeWhQiiftlOCkgi.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Mission: Impossible - Fallout
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2018
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Adventure & Action & Thriller
      </p>
    </article>
  </div>,
  <div
    data-testid="card_movie"
  >
    <article
      class="main-module__movieItem___2LLLX"
    >
      <div
        class="main-module__menuContainer___ZkGch"
      >
        <div
          class="main-module__menuButton___6XkhL"
          data-cy="movie_menu_button"
          data-testid="movie_menu_button"
        >
          ...
        </div>
      </div>
      <img
        alt="Jurassic World: Fallen Kingdom"
        class="main-module__movieImage___39dYg"
        data-testid="card_movie_image"
        src="https://image.tmdb.org/t/p/w500/ln6d5Okr6VK5vfQVobJTiYxeD0l.jpg"
      />
      <div
        class="main-module__titleContainer___3i1_O"
      >
        <p
          class="main-module__movieTitle___3k_wC"
          data-cy="card_movie_title"
        >
          Jurassic World: Fallen Kingdom
        </p>
        <p
          class="main-module__movieDate___thIkl"
        >
          2018
        </p>
      </div>
      <p
        class="main-module__genres___qwVlZ"
      >
        Action & Adventure & Drama & Science Fiction & Thriller
      </p>
    </article>
  </div>,
]
`;
